### PR TITLE
raftstore: add cache for snapshot

### PR DESCRIPTION
**Pull request by 5kbpers**
## What have you changed? (mandatory)

This PR removes the local reader thread and caches snapshot before writing to db.

## What are the type of the changes? (mandatory)

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
- Improvement (change which is an improvement to an existing feature)

### CODE REVIEW

File: components/engine/src/rocks/snapshot.rs [hunk 1]
```diff
@@ -60,6 +60,14 @@ impl Snapshot {
         }
         Ok(DBIterator::new_cf(Arc::clone(&self.db), handle, opt))
     }
+
+    pub fn get_sequence_number(&self) -> u64 {
```
Comment (siddontang) [inline components/engine/src/rocks/snapshot.rs]: em, I remember Snapshot now has a get sequence number API, you don't need to call from FFI

Comment (5kbpers) [reply components/engine/src/rocks/snapshot.rs]: The Snapshot struct there doesn't use the Snapshot struct of our rocksdb-rust crate, so I have to add a method there.
I think it needs to be refactored by using the Snapshot struct of rocksdb-rust directly.

Comment (5kbpers) [reply components/engine/src/rocks/snapshot.rs]: Or we can add the `get_sequence_number` method to the `UnsafeSnap` struct.

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -2319,11 +2327,21 @@ impl ReadExecutor {
         if self.snapshot.is_some() {
             return;
         }
-        let engine = self.engine.clone();
-        self.snapshot = Some(Snapshot::new(engine).into_sync());
-        // Reading current timespec after snapshot, in case we do not
-        // expire lease in time.
-        atomic::fence(atomic::Ordering::Release);
+
+        loop {
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: do we really need `loop` here?

seems if we get option_snapshot after write and the option_snapshot is not `None`, we can use this option_snapshot directly.

File: src/server/node.rs [hunk 3]
```diff
@@ -114,6 +113,7 @@ where
     pub fn start<T>(
         &mut self,
         engines: Engines,
+        engine_snapshot: Arc<RwLock<Option<SyncSnapshot>>>,
```
Comment (siddontang) [inline src/server/node.rs]: can we move this to Engines, seem it is weird to pass an Engines and an engine snapshot together.

**Comment by overvenus** [comment]
PTAL @5kbpers  , should we close this PR?

**Comment by overvenus** [comment]
This PR is being closed because it is inactive.

**Pull request closed without merging**